core/src/sigverifier/bls_sigverifier.rs

```diff
@@ -25,21 +39,224 @@ impl SigVerifier for BLSSigVerifier {
         packet_batches: Vec<PacketBatch>,
     ) -> Result<(), SigVerifyServiceError<Self::SendType>> {
         // TODO(wen): just a placeholder without any batching.
-        if let Some(sender) = &self.sender {
-            packet_batches.iter().for_each(|batch| {
-                batch.iter().for_each(|packet| {
-                    if let Ok(message) = packet.deserialize_slice::<BLSMessage, _>(..) {
-                        sender.send(message).unwrap();
+        let mut packet_received = 0;
+        let mut bls_messages_sent = 0;
+        let mut bls_messages_malformed = 0;
+        let mut sent_failed = 0;
+        packet_batches.iter().for_each(|batch| {
+            batch.iter().for_each(|packet| {
+                packet_received += 1;
+                match packet.deserialize_slice::<BLSMessage, _>(..) {
+                    Ok(message) => match self.sender.try_send(message) {
+                        Ok(()) => {
+                            bls_messages_sent += 1;
+                        }
+                        Err(TrySendError::Full(_)) => {
+                            warn!("BLS message channel is full, dropping message");
+                            sent_failed += 1;
+                        }
+                        Err(TrySendError::Disconnected(_)) => {
+                            // There is no hope to recover, restart the validator.
+                            panic!("BLS message channel is disconnected");
+                        }
+                    },
+                    Err(e) => {
+                        trace!("Failed to deserialize BLS message: {}", e);
+                        bls_messages_malformed += 1;
                     }
-                });
+                }
             });
+        });
+        if packet_received > 0 {
+            self.stats.packets_received += packet_received;
+            self.stats.bls_messages_sent += bls_messages_sent;
+            self.stats.bls_messages_malformed += bls_messages_malformed;
+            self.stats.sent_failed += sent_failed;
         }
+        // We don't need lock on stats because stats are read and write in a single thread.
+        self.report_stats();
         Ok(())
     }
 }
 
 impl BLSSigVerifier {
-    pub fn new(sender: Option<Sender<BLSMessage>>) -> Self {
-        Self { sender }
+    fn report_stats(&mut self) {
```

Comment:
I think we conventionally use something like `maybe_report_stats` when there are conditionals inside. But I'm okay with whatever